examples/transactions.c

```diff
@@ -321,112 +293,119 @@ static void commit_transaction_and_start_new (struct state *state) {
-                        struct state *state = calloc(1, sizeof(*state));
-
-                        state->producer = create_transactional_producer(rktpar);
-                        state->rktpar = rktpar;
-                        rktpar->opaque = state;
-                        state->last_commit = time(NULL);
-
-                        fprintf(stdout,
-                                " %s [%d] with transactional producer %s\n",
-                                rktpar->topic, rktpar->partition,
-                                rd_kafka_name(state->producer));
-                }
-
-                /* Let the consumer know the rebalance has been handled
-                 * by calling assign.
-                 * This will also tell the consumer to start fetching messages
-                 * for the assigned partitions. */
-                rd_kafka_assign(rk, partitions);
+                fprintf(stdout,
+                        "Consumer group rebalanced: "
+                        "%d new partition(s) assigned\n",
+                        partitions->cnt);
+
+                /* Start fetching messages for the assigned partitions
+                 * and add them to the consumer's local assignment. */
+                error = rd_kafka_incremental_assign(consumer, partitions);
+                if (error)
+                        fatal_error("Incremental assign failed", error);
                 break;
 
         case RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS:
-                fprintf(stdout,
-                        "Consumer group rebalanced, assignment revoked\n");
+                if (!rd_kafka_assignment_lost(consumer)) {
+                        fprintf(stdout,
+                                "Consumer group rebalanced: assignment lost: "
+                                "aborting current transaction\n");
 
-                /* Abort the current transactions and destroy all producers */
-                for (i = 0 ; i < assigned_partitions->cnt ; i++) {
-                        /* Store the partition-to-producer mapping
-                         * in the partition's opaque field. */
-                        struct state *state = (struct state *)
-                                assigned_partitions->elems[i].opaque;
+                        error = rd_kafka_abort_transaction(producer, -1);
+                        if (error)
+                                fatal_error("Failed to abort transaction",
+                                            error);
+                } else {
+                        fprintf(stdout,
+                                "Consumer group rebalanced: %d partition(s) "
+                                "revoked: committing current transaction\n",
+                                partitions->cnt);
 
-                        destroy_transactional_producer(state->producer);
-                        free(state);
+                        commit_transaction(consumer, producer);
```

Comment:
don't you need to send offsets to transaction here as well?

Reply:
commit_transaction() does that

Reply:
ahh, whoops sorry